### PR TITLE
subsys: kernel_shell: extend thread info

### PR DESCRIPTION
**Pull request by pavlohamov**
1) Dump time sinse last scheduler call
Could be handy for tickless kernel debug.
Will indicate that no rtc irq is called

2) Dump current timeout of each thread
Could be used to find yout when thread will wake up

3) Dump human friendly thread state

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

### CODE REVIEW

File: subsys/shell/modules/kernel_service.c
Comment (nordic-krch) [inline subsys/shell/modules/kernel_service.c]: that would look nicer if you turn it to a function (e.g. `thread_state_to_str()`)

File: subsys/shell/modules/kernel_service.c
Comment (nordic-krch) [inline subsys/shell/modules/kernel_service.c]: `timeout:`

File: subsys/shell/modules/kernel_service.c
Comment (nordic-krch) [inline subsys/shell/modules/kernel_service.c]: please check `shell_print` macro. Could be used here (it prints in normal and adds new line).

File: subsys/shell/modules/kernel_service.c
Comment (nordic-krch) [inline subsys/shell/modules/kernel_service.c]: if  you are willing to do a bit of cleanup you could create local shell variable instead of casting it in each call and use `shell_print` which looks more compact.

**Review by nordic-krch — CHANGES_REQUESTED** [review]

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: yeah, make sense. Thank, you for this idea. Done

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: Thats odd. I even missed that it's **f**printf. whole file is using the fprintf/ maybe i will use proper one. Does it make sense to update whole file in separate PR?

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: Looks nicer. Thanks, done.

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: Seems like i'm too lazy to create a separate PR. Done in the same

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: Oups. Thank you for noticing this typo. Done

**Comment by pavlohamov** [comment]
@nordic-krch Please take a look. Verified on stm32_min_dev and nrf52832 (custom board)

File: subsys/shell/modules/kernel_service.c
Comment (nordic-krch) [inline subsys/shell/modules/kernel_service.c]: i was more thinking about `const char *thread_state_to_str(u8_t state)`. That could even be moved at some point to some kernel header (e.g. `kernel_structs.h` where flags are defined) since it would be generic and might be reused by logger, printk or shell.

File: subsys/shell/modules/kernel_service.c
Comment (nordic-krch) [inline subsys/shell/modules/kernel_service.c]: please fix those shell_prints which will now fit in 80 lines.

**Review by nordic-krch — CHANGES_REQUESTED** [review]

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: Mmm, return string is non-const. Any combination of flags is possible. Thus string must be given as parameter, which makes usage a bit complicated.

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: Done

File: subsys/shell/modules/kernel_service.c
Comment (nordic-krch) [reply subsys/shell/modules/kernel_service.c]: >  Any combination of flags is possible

I see in fact that in some places there are `thread->base.thread_state |= _THREAD_DEAD` and `thread->base.thread_state |= _THREAD_ABORTING;` assignments but i wonder if that is correct. State should be exclusive.
@andrewboie  @andyross can you help with that? Is this or assignment a bug or a feature?

File: subsys/shell/modules/kernel_service.c
Comment (jakub-uC) [inline subsys/shell/modules/kernel_service.c]: Do you need `\n` at the end?

File: subsys/shell/modules/kernel_service.c
Comment (jakub-uC) [inline subsys/shell/modules/kernel_service.c]: Just a hint. If you do not need fixed width `-10` you can use variable length with `*`.

**Review by jakub-uC — APPROVED** [review]

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: @nordic-krch Seems like i found one place where state is being set as a combination of flags.
https://github.com/zephyrproject-rtos/zephyr/blob/11e91a83c1f6923a0078834404e238e5b7df3693/arch/x86/core/ia32/float.c#L211
defined at 
https://github.com/zephyrproject-rtos/zephyr/blob/11e91a83c1f6923a0078834404e238e5b7df3693/arch/x86/include/ia32/kernel_arch_data.h#L62

What seems to be kinda strange:
```
/* executing context is interrupt handler */
#define _INT_ACTIVE (1 << 7)
/* executing context is exception handler */
#define _EXC_ACTIVE (1 << 6)
```
```
/* Thread is present in the ready queue */
#define _THREAD_QUEUED (BIT(6))
```
ia32 _EXC_ACTIVE overrides the _THREAD_QUEUED flag. Don't you know is it normal?

**Comment by pavlohamov** [comment]
@nordic-krch i've added the "k_thread_state_str" function. Could you be so kind to take a look?

File: kernel/thread.c
Comment (nordic-krch) [inline kernel/thread.c]: since there is a chance that thread will be in multiple (!) states I would not assert. Just return `unknown`.

File: subsys/shell/modules/kernel_service.c
Comment (pavlohamov) [reply subsys/shell/modules/kernel_service.c]: Hm, on other hand Exception or Interrupt execution on x86 is interrupt normal program execution, so we can assume that both flags are used to indicate t hat we're in this context. Seems like only _EXC_ACTIVE == _THREAD_QUEUED  - is suspicious.

File: kernel/thread.c
Comment (pavlohamov) [reply kernel/thread.c]: I'm thinking that your first idea that state is exclusive seems to be valid. But right now i agree that this assert must be removed. Done

If x86 states must differ from the generic thread states
```
#define _INT_ACTIVE (1 << 7)
#define _EXC_ACTIVE (1 << 6)
```
They must become BIT(7) & BIT(8) which overfill the 8 bit value. But if we assume that states are exclusive, bitmask could be replaced with simple enum.

Comment (nordic-krch) [reply kernel/thread.c]: current implementation of scheduler does not perform atomic transitions from one state to another, e.g. going to pending state is done in steps: remove from ready queue and unset ready state, go to pend state. In that example for a moment thread is in dummy state. I don't know if it may lead to some issues apart from logging. Maintainers have been pinged here so let them decide.

**Review by nordic-krch — APPROVED** [review]

**Comment by pavlohamov** [comment]
@andyross @andrewboie Could you be so kind to take a look?

**Comment by andrewboie** [comment]
> @andyross @andrewboie Could you be so kind to take a look?

@andyross would be the best person to comment, my knowledge of the scheduler is somewhat superficial.

However it does appear that state transitions are not atomic as @nordic-krch says so I think you should remove that assertion, and consider iterating over bits in the state field instead of fixed values when dumping stuff out

File: kernel/thread.c
Comment (andrewboie) [inline kernel/thread.c]: think this is fundamentally flawed, although it might work most of the time.
better to examine each bit at a time.

Comment (nordic-krch) [reply kernel/thread.c]: but won't it create more confusion if by any chance shell output indicates that thread is in two states? Isn't it better to print `unknown` or `in transition` for that very unlikely (to be read out in low priority thread) transient state? at least state naming suggests that states are exclusive.

Comment (pavlohamov) [reply kernel/thread.c]: https://github.com/zephyrproject-rtos/zephyr/blob/77db273f6f84b3d94d34b5e72070ec6bc31e41cf/lib/cmsis_rtos_v2/thread.c#L302
```	
        case _THREAD_DUMMY:
		state = osThreadError;
		break;
	case _THREAD_PRESTART:
		state = osThreadInactive;
		break;
```
Treats thread state as exclusive. 
Suppose, we need to get rid of this implicity in thread states (smth. like simple comment in header file)

**Comment by carlescufi** [comment]
@andyross @andrewboie could you please take another look before we merge?

File: kernel/thread.c
Comment (carlescufi) [reply kernel/thread.c]: Since this is the main contention point, can @andyross please weigh in?

**Comment by pavlohamov** [comment]
@andyross kind reminder. Could you be so kind to take a look?

File: kernel/thread.c
Comment (andrewboie) [reply kernel/thread.c]: fine w/this as-is

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**
